python/ql/src/experimental/semmle/python/frameworks/XML.qll

```diff
@@ -0,0 +1,196 @@
+    }
+  }
+
+  /** Gets a reference to `lxml`. */
+  private API::Node lxml() { result = API::moduleImport("lxml") }
+
+  /** Gets a reference to `lxml.etree`. */
+  private API::Node lxmlEtree() { result = lxml().getMember("etree") }
+
+  /**
+   * Gets a call to `lxml.etree.XMLParser` or `lxml.etree.get_default_parser` and `mayBeDangerous()`
+   * identifies whether the argument `no_network` is set to `False` or the arguments `huge_tree`
+   * or `resolve_entities` are set to True.
+   */
+  private class LXMLParser extends DataFlow::CallCfgNode, XMLParser::Range {
+    LXMLParser() { this = lxmlEtree().getMember(["XMLParser", "get_default_parser"]).getACall() }
+
+    override DataFlow::Node getAnInput() { none() }
+
+    override predicate mayBeDangerous() {
+      not exists(this.getArgByName(_)) or
+      DataFlow::localFlow(DataFlow::exprNode(any(False falseName)), this.getArgByName("no_network")) or
+      DataFlow::localFlow(DataFlow::exprNode(any(True trueName)),
+        this.getArgByName(["huge_tree", "resolve_entities"]))
+    }
+  }
+
+  /**
+   * Gets a call to `lxml.etree.fromstring`, `xml.etree.fromstringlist`,
+   * `xml.etree.XML` or `xml.etree.parse`.
+   *
+   * Given the following example:
+   *
+   * ```py
+   * parser = lxml.etree.XMLParser()
+   * parsed_xml = lxml.etree.fromstring(xml_content, parser=parser).text
+   * ```
+   *
+   * `this` would be `lxml.etree.fromstring(xml_content, parser=parser)`
+   * and `xml_content` would be the result of `getAnInput()`.
+   */
+  private class LXMLParsing extends DataFlow::CallCfgNode, XMLParsing::Range {
+    LXMLParsing() {
+      this = lxmlEtree().getMember(["fromstring", "fromstringlist", "XML", "parse"]).getACall()
+    }
+
+    override DataFlow::Node getAnInput() { result = this.getArg(0) }
+
+    override predicate mayBeDangerous() {
+      exists(XMLParser xmlParser |
+        xmlParser.mayBeDangerous() and this.getArgByName("parser").getALocalSource() = xmlParser
+      )
+      or
+      not exists(this.getArgByName("parser"))
+    }
+  }
+
+  /** Gets a reference to the `xmltodict` module. */
+  private API::Node xmltodict() { result = API::moduleImport("xmltodict") }
+
+  /**
+   * Gets a call to `xmltodict.parse` and `mayBeDangerous()` identifies
+   * whether the argument `disable_entities` is set to `False`.
+   */
+  private class XMLtoDictParsing extends DataFlow::CallCfgNode, XMLParsing::Range {
+    XMLtoDictParsing() { this = xmltodict().getMember("parse").getACall() }
+
+    override DataFlow::Node getAnInput() { result = this.getArg(0) }
+
+    override predicate mayBeDangerous() {
+      DataFlow::localFlow(DataFlow::exprNode(any(False falseName)),
+        this.getArgByName("disable_entities"))
+    }
+  }
+
+  /** Gets a reference to `xml.dom.minidom` or `xml.dom.pulldom`. */
+  private API::Node xmlDom() { result = xml().getMember("dom").getMember(["mini", "pull"] + "dom") }
```

Comment:
It's clever you _can_ do `["mini", "pull"] + "dom"`, but I would rather like this for readability.

```suggestion
  private API::Node xmlDom() { result = xml().getMember("dom").getMember(["minidom", "pulldom"]) }
```